spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/query/QueryValidator.java

```diff
@@ -0,0 +1,116 @@
+/*
+ *  Copyright (c) 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.spi.query;
+
+import org.eclipse.dataspaceconnector.common.reflection.ReflectionUtil;
+import org.eclipse.dataspaceconnector.spi.result.Result;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.ParameterizedType;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+
+/**
+ * Validates that a particular query is valid, i.e. contains a left-hand operand, that conforms to the canonical
+ * format.
+ */
+public class QueryValidator {
+    private final Class<?> canonicalType;
+    private final Map<Class<?>, List<Class<?>>> subtypeMap;
+
+    /**
+     * Constructs a new QueryValidator instance.
+     *
+     * @param canonicalType The Java class of the object to validate against.
+     * @param typeHierarchyMap Contains mapping from superclass to list of subclasses. Every superclass must be
+     *         represented as separate entry in the map, even if it is also a subclass of another.
+     */
+    public QueryValidator(Class<?> canonicalType, Map<Class<?>, List<Class<?>>> typeHierarchyMap) {
+        this.canonicalType = canonicalType;
+        subtypeMap = typeHierarchyMap;
+    }
+
+    public QueryValidator(Class<?> canonicalType) {
+        this(canonicalType, null);
+    }
+
+    /**
+     * Validates a {@link QuerySpec} whether it conforms to a particular schema (e.g. a Java class) or not
+     */
+    public Result<Void> validate(QuerySpec query) {
+        return query.getFilterExpression().stream()
+                .map(Criterion::getOperandLeft)
+                .map(Object::toString)
+                .map(this::isValid)
+                .reduce((res1, res2) -> {
+                    res1.getFailureMessages().addAll(res2.getFailureMessages());
```

Comment:
This will throw an NPE if the filter expression list contains at least one successful `Result`.
I would avoid this implementing it in a functional way: adding the `and` operator on the `Result` type, something like:
```java
    public Result<Void> and(Result<Void> other) {
        if (succeeded() && other.succeeded()) {
            return Result.success();
        } else {
            var messages = new ArrayList<String>();
            messages.addAll(Optional.ofNullable(this.getFailure()).map(Failure::getMessages).orElse(Collections.emptyList()));
            messages.addAll(Optional.ofNullable(other.getFailure()).map(Failure::getMessages).orElse(Collections.emptyList()));
            return Result.failure(messages);
        }
    }
```

so this lambda could be written just as:
```java
.reduce(Result::and)
```